File: backend/scripts/populate_cuisines/populate_cuisines.go
package main

import (
	"database/sql"
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/google/uuid"
	"github.com/joho/godotenv"
	_ "github.com/lib/pq"
)

// DBConfig holds the database configuration
type DBConfig struct {
	DBName   string
	User     string
	Password string
	Host     string
	Port     string
}

// getParentID retrieves the ID of the parent row by its name
func getParentID(db *sql.DB, parentName, tableName string) (sql.NullString, error) {
	query := fmt.Sprintf("SELECT id FROM %s WHERE name = $1", tableName)
	var parentID sql.NullString
	err := db.QueryRow(query, parentName).Scan(&parentID)
	if err == sql.ErrNoRows {
		return parentID, nil
	}
	return parentID, err
}

// importCSVToPostgres imports data from a CSV file into a PostgreSQL table
func importCSVToPostgres(csvFilePath string, dbConfig DBConfig, tableName string) error {
	connStr := fmt.Sprintf("dbname=%s user=%s password=%s host=%s port=%s sslmode=disable",
		dbConfig.DBName, dbConfig.User, dbConfig.Password, dbConfig.Host, dbConfig.Port)
	db, err := sql.Open("postgres", connStr)
	if err != nil {
		return err
	}
	defer db.Close()

	file, err := os.Open(csvFilePath)
	if err != nil {
		return err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return err
	}

	for i, row := range records {
		if i == 0 {
			continue // Skip header row
		}
		rowID := uuid.New().String()
		name := row[0]
		parentName := row[1]
		createdAt := time.Now()
		updatedAt := createdAt

		var parentID sql.NullString
		if parentName != "" {
			parentID, err = getParentID(db, parentName, tableName)
			if err != nil {
				return err
			}
		}

		insertQuery := fmt.Sprintf(
			"INSERT INTO %s (id, name, parent_id, created_at, updated_at) VALUES ($1, $2, $3, $4, $5)",
			tableName,
		)
		_, err = db.Exec(insertQuery, rowID, name, parentID, createdAt, updatedAt)
		if err != nil {
			return err
		}
	}

	return nil
}

func main() {
	if err := godotenv.Load(); err != nil && !os.IsNotExist(err) {
		log.Fatal("error loading env file: database")
	}

	dbConfig := DBConfig{
		DBName:   os.Getenv("DB_NAME"),
		User:     os.Getenv("DB_USER"),
		Password: os.Getenv("DB_PASSWORD"),
		Host:     os.Getenv("DB_HOST"),
		Port:     os.Getenv("DB_PORT"),
	}

	csvFilePath := "data/cuisines.csv"
	tableName := "cuisines"

	err := importCSVToPostgres(csvFilePath, dbConfig, tableName)
	if err != nil {
		log.Fatalf("error importing CSV to PostgreSQL: %v", err)
	}

	fmt.Println("cuisines data successfully imported!")
}

File: backend/scripts/populate_cuisines/populate_cuisines.py
import csv
import os
import uuid
from datetime import datetime

import psycopg2
from dotenv import load_dotenv
from psycopg2 import sql, connect


def get_parent_id(cursor, parent_name, table_name):
    query = sql.SQL("SELECT id FROM {table} WHERE name = %s").format(
        table=sql.Identifier(table_name)
    )
    cursor.execute(query, (parent_name,))
    result = cursor.fetchone()
    return result[0] if result else None


def import_csv_to_postgres(csv_file_path, db_config, table_name):
    # Connect to PostgreSQL database
    conn = connect(
        dbname=db_config["dbname"],
        user=db_config["user"],
        password=db_config["password"],
        host=db_config["host"],
        port=db_config["port"],
    )
    cursor = conn.cursor()

    # Open CSV file
    with open(csv_file_path, mode="r") as file:
        reader = csv.DictReader(file)

        # Iterate over CSV rows and insert into database
        for row in reader:
            row_id = str(uuid.uuid4())  # Generate a unique UUID
            name = row["name"]
            parent_name = row["parent"]
            created_at = datetime.now()
            updated_at = created_at

            parent_id = None
            if parent_name:
                parent_id = get_parent_id(cursor, parent_name, table_name)

            insert_query = sql.SQL(
                """
                INSERT INTO {table} (id, name, parent_id, created_at, updated_at)
                VALUES (%s, %s, %s, %s, %s)
            """
            ).format(table=sql.Identifier(table_name))
            cursor.execute(
                insert_query, (row_id, name, parent_id, created_at, updated_at)
            )

    # Commit changes and close the connection
    conn.commit()
    cursor.close()
    conn.close()


if __name__ == "__main__":
    load_dotenv()  # Load environment variables from .env file

    db_config = {
        "dbname": os.getenv("DB_NAME"),
        "user": os.getenv("DB_USER"),
        "password": os.getenv("DB_PASSWORD"),
        "host": os.getenv("DB_HOST"),
        "port": os.getenv("DB_PORT"),
    }

    csv_file_path = "scripts/data/cuisines.csv"
    table_name = "cuisines"

    import_csv_to_postgres(csv_file_path, db_config, table_name)
